### PR TITLE
add new feature AHP_depth_from_AP_begin_voltage (and slow version)

### PR DESCRIPTION
**Pull request by arnaudon**

### CODE REVIEW

File: efel/cppcore/LibV5.cpp [hunk 1]
```diff
@@ -3993,3 +3993,83 @@ int LibV5::time_to_interburst_min(mapStr2intVec& IntFeatureData,
...
+                     mapStr2doubleVec& DoubleFeatureData,
+                     mapStr2Str& StringData) {
+  int retval;
+  int nsize;
+  retval = CheckInMap(DoubleFeatureData, StringData, "AHP_depth_from_AP_begin_voltage", nsize);
+  if (retval) {
+    return nsize;
+  }
+
+  vector<double> voltagebase;
```
Comment (AurelienJaquier) [inline efel/cppcore/LibV5.cpp:4018]: You could maybe rename this feature for clarity? something like apbeginvoltage

File: efel/cppcore/LibV5.cpp [hunk 2]
```diff
@@ -3993,3 +3993,83 @@ int LibV5::time_to_interburst_min(mapStr2intVec& IntFeatureData,
...
+    setVec(DoubleFeatureData, StringData, "AHP_depth_from_AP_begin_voltage", ahpdepth);
+  }
+  return retval;
+}
+// end of AHP_depth_from_AP_begin_voltage
+
+
+// *** AHP_depth_slow_from_AP_begin_voltage ***
+static int __AHP_depth_slow_from_AP_begin_voltage(const vector<double>& ap_begin_voltage,
+                       const vector<double>& minahpvalues,
```
Comment (AurelienJaquier) [inline efel/cppcore/LibV5.cpp:4039]: maybe also rename minahpvalues into ahpdepthabsslow for clarity

File: efel/cppcore/LibV5.cpp [hunk 3]
```diff
@@ -3993,3 +3993,83 @@ int LibV5::time_to_interburst_min(mapStr2intVec& IntFeatureData,
...
+                     mapStr2doubleVec& DoubleFeatureData,
+                     mapStr2Str& StringData) {
+  int retval;
+  int nsize;
+  retval = CheckInMap(DoubleFeatureData, StringData, "AHP_depth_slow_from_AP_begin_voltage", nsize);
+  if (retval) {
+    return nsize;
+  }
+
+  vector<double> voltagebase;
```
Comment (AurelienJaquier) [inline efel/cppcore/LibV5.cpp:4057]: same comment here

**Review by AurelienJaquier — COMMENTED** [review]
Overall looks good. You should still add your features to tests/featurenames.json and testdata/allfeatures/expectedresults.json for the tests to pass ;)

File: efel/tests/test_basic.py [hunk 4]
```diff
@@ -3176,6 +3176,75 @@ def test_AHP_depth_slow():
...
+
+    AHP_depth_from_AP_begin_voltage = feature_values[0][
+        "AHP_depth_from_AP_begin_voltage"
+    ]
+    AHP_depth_abs = feature_values[0]["AHP_depth_abs"]
+    AP_begin_voltage = feature_values[0]["AP_begin_voltage"]
+
+    # we do this to mimic the cpp implementation, sometimes different AP are
+    # picked up  but as we consider averages, it's ok. To fix at some point.
+    expected = [v - ahp for v, ahp in zip(AP_begin_voltage, AHP_depth_abs)]
```
Comment (AurelienJaquier) [inline efel/tests/test_basic.py:3205]: Using numpy, we could just do `expected = AP_begin_voltage - AHP_depth_abs` I believe.

Comment (arnaudon) [reply efel/tests/test_basic.py:3205]: no, this does not work, I explained in the comment above why, but I'm really unsure on how to fix it.

Comment (AurelienJaquier) [reply efel/tests/test_basic.py:3205]: I don't get your comment. What do you mean by 'sometimes, differnet APs are picked up'. That sounds pretty bad

**Comment by arnaudon** [comment]
@AurelienJaquier I found the issue here, it was that the implementation of AHP_depth_slow was ignoring the first and last APs, I'm not sure why at all. Just playing with indices now makes more sense, as it returns the values for the same AP as AHP_depth or AP_begin_voltage.

**Comment by AurelienJaquier** [comment]
I don't think you should change AHP_depth_slow behavior. Ignoring the first peak is not a bug, it is a feature. I think the reason is because we can encounter traces with an early peak before even the stimulus start.

If you want AHP_depth_slow to be able to also catch the 1st ISI, please make it dependent on the i`gnore_first_ISI` setting.

**Comment by arnaudon** [comment]
mmh I'm not sure this is the reason, as this feature start detecting APs after stim_start, no? it should catch the same AP as AHP_depth IMO, if we set the delay to 0ms, then AHP_depth should be == to AHP_depth_slow, no?

**Comment by AurelienJaquier** [comment]
It isn't bound by stim start. But yes, it would be more coherent if AHP_depth_slow and AHP_depth followed the same logic

**Comment by arnaudon** [comment]
Ah the peak_indices is not between stim_start and stim_end? it should no? what is the reason for that? do you know?

**Comment by AurelienJaquier** [comment]
I would guess it's because for some cases we want to know about peaks outside stimulus, like rebound, etc.

**Comment by arnaudon** [comment]
I tried to do AHP_depth_abs_slow more like AHP_depth_abs, to be continued

**Pull request closed without merging**